ts/components/registration/stages/RestoreAccount.tsx

```diff
@@ -0,0 +1,321 @@
+import { AccountDetails } from './CreateAccount';
+
+type AccountRestoreDetails = AccountDetails & { dispatch: Dispatch; abortSignal?: AbortSignal };
+
+export async function finishRestore(pubkey: string, displayName: string) {
+  await setSignWithRecoveryPhrase(true);
+  await registrationDone(pubkey, displayName);
+
+  window.log.debug(`WIP: [onboarding] restore account: logging in for ${displayName}`);
+  trigger('openInbox');
+}
+
+/**
+ * This will try to sign in with the user recovery password.
+ * If no ConfigurationMessage is received within ONBOARDING_RECOVERY_TIMEOUT, the user will be asked to enter a display name.
+ */
+async function signInAndFetchDisplayName(args: AccountRestoreDetails) {
+  const { recoveryPassword, dispatch, abortSignal } = args;
+
+  try {
+    await resetRegistration();
+    const promiseLink = signInByLinkingDevice(recoveryPassword, 'english', abortSignal);
+    const promiseWait = PromiseUtils.waitForTask(done => {
+      window.Whisper.events.on(
+        'configurationMessageReceived',
+        async (ourPubkey: string, displayName: string) => {
+          window.Whisper.events.off('configurationMessageReceived');
+          await setSignInByLinking(false);
+          dispatch(setHexGeneratedPubKey(ourPubkey));
+          dispatch(setDisplayName(displayName));
+          dispatch(setAccountRestorationStep(AccountRestoration.Finishing));
+          done(displayName);
+        }
+      );
+    }, ONBOARDING_TIMES.RECOVERY_TIMEOUT);
+
+    await Promise.all([promiseLink, promiseWait]);
+  } catch (e) {
+    await resetRegistration();
+    throw e;
+  }
+}
+
+/**
+ * Sign in/restore from seed.
+ * Ask for a display name, as we will drop incoming ConfigurationMessages if any are saved on the swarm.
+ * We will handle a ConfigurationMessage
+ */
+async function signInWithNewDisplayName(args: AccountRestoreDetails) {
+  const { displayName, recoveryPassword, dispatch } = args;
+
+  try {
+    const validDisplayName = displayNameIsValid(displayName);
+
+    await resetRegistration();
+    await registerSingleDevice(
+      recoveryPassword,
+      'english',
+      validDisplayName,
+      async (pubkey: string) => {
+        dispatch(setHexGeneratedPubKey(pubkey));
+        dispatch(setDisplayName(validDisplayName));
+        await finishRestore(pubkey, validDisplayName);
+      }
+    );
+  } catch (e) {
+    await resetRegistration();
+    throw e;
+  }
+}
+
+export const RestoreAccount = () => {
+  const step = useOnboardAccountRestorationStep();
+  const recoveryPassword = useRecoveryPassword();
+  const recoveryPasswordError = useRecoveryPasswordError();
+  const ourPubkey = useOnboardHexGeneratedPubKey();
+  const displayName = useDisplayName();
+  const displayNameError = useDisplayNameError();
+  const progress = useProgress();
+
+  const dispatch = useDispatch();
+
+  useRecoveryProgressEffect({
+    step,
+    progress,
+    setProgress,
+    ourPubkey,
+    displayName,
+  });
```

Comment:
you could move some of those hooks call in the `useRecoveryProgressEffect` and return them from the hook rather than passing them. It's a bit weird to pass something to the hook which is changed by the hook.

So something like

```ts

const {step, progress} =  useRecoveryProgressEffect({
//    setProgress, it's just an import that you can do in the hook itself too
    ourPubkey,
    displayName,
  });